subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1235 @@
+				  struct http_client_ctx *ctx_client, int client_index)
+{
+	ctx_client->server_state = HTTP_SERVER_FRAME_GOAWAY_STATE;
+
+	return 0;
+}
+
+int enter_http_http_done_state(struct http_server_ctx *ctx_server,
+			       struct http_client_ctx *ctx_client, int client_index)
+{
+	ctx_client->server_state = HTTP_SERVER_DONE_STATE;
+
+	return 0;
+}
+
+int handle_http_preface(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_SERVER_PREFACE_STATE.");
+	if (ctx_client->offset < strlen(preface)) {
+		/* We don't have full preface yet, get more data. */
+		return -EAGAIN;
+	}
+
+	if (strncmp(ctx_client->buffer, preface, strlen(preface)) != 0) {
+		ctx_client->server_state = HTTP_SERVER_REQUEST_STATE;
+	} else {
+		ctx_client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+
+		ctx_client->offset -= strlen(preface);
+		memmove(ctx_client->buffer, ctx_client->buffer + strlen(preface),
+			ctx_client->offset);
+	}
+
+	return 0;
+}
+
+int handle_http1_request(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+			 int client_index)
+{
+	LOG_DBG("HTTP_SERVER_REQUEST.");
+	const char *data;
+	int len;
+	int total_received = 0;
+	int offset = 0;
+
+	http_parser_init(&parser, HTTP_REQUEST);
+	http_parser_settings_init(&parserSettings);
+	parserSettings.on_header_field = on_header_field;
+	parserSettings.on_url = on_url;
+	http_parser_execute(&parser, &parserSettings, ctx_client->buffer + offset,
+			    ctx_client->offset);
+
+	total_received += ctx_client->offset;
+	offset += ctx_client->offset;
+
+	if (offset >= HTTP_SERVER_MAX_RESPONSE_SIZE) {
+		offset = 0;
+	}
+
+	if (has_upgrade_header == false) {
+		const char *response = "HTTP/1.1 101 Switching Protocols\r\n"
+				       "Connection: Upgrade\r\n"
+				       "Upgrade: h2c\r\n"
+				       "\r\n";
+		if (sendall(ctx_client->client_fd, response, strlen(response)) < 0) {
+			close_client_connection(ctx_server, client_index);
+		}
+
+		memset(ctx_client->buffer, 0, sizeof(ctx_client->buffer));
+		ctx_client->offset = 0;
+		ctx_client->server_state = HTTP_SERVER_PREFACE_STATE;
+
+	} else {
+		data = content_200;
+		len = sizeof(content_200);
+
+		enum http_method method = parser.method;
+		const char *method_str = http_method_str(method);
+
+		LOG_DBG("HTTP Method: %s", method_str);
+
+		if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/results", 8) == 0) {
+			handle_get_request(ctx_server, ctx_client->client_fd);
+		} else if (strncmp(method_str, "POST", 4) == 0 &&
+			   strncmp(url_buffer, "/add", 4) == 0) {
+			char *json_start = strstr(ctx_client->buffer, "\r\n\r\n");
+
+			if (json_start) {
+				handle_post_request(ctx_server, json_start);
+			}
+
+			char response[] = "HTTP/1.1 200 OK\r\nContent-Type: "
+					  "text/plain\r\n\r\nAdded successfully.\n";
+			sendall(ctx_client->client_fd, response, strlen(response));
+		} else if (strncmp(method_str, "GET", 3) == 0 && strncmp(url_buffer, "/", 1) == 0) {
```

Comment:
Again, why is an application-specific resource ("/") hard-coded in the http_server library code?